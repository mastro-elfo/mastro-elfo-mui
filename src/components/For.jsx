import React, { Fragment } from "react";
import PropTypes from "prop-types";

/**
 * For `each` item apply `map` and create `Component`.
 *
 * A `key` property is given by the array index of the element. This can be overwritten by `map`.
 *
 * @param {Array}       [each=[]]  Array to loop
 * @param {func}        [map=(x)=>x]  Apply map to each item
 * @param {elementType} [Component] (Required) Create a component for each item
 * @param {object}      [ComponentProps={}] Common properties for each item
 * @param {elementType} [Container=Fragment] List container
 * @param {object}      [ContainerProps={}] Properties for the Container component
 * @constructor
 */
export default function For({
  // Items to loop
  each = [],
  // Map items, default identity
  map = (x) => x,
  // Component and Container elements and props
  Component,
  ComponentProps = {},
  Container = Fragment,
  ContainerProps = {},
}) {
  // If `each` is not a valid array
  if (!each || !each.map) return <Container {...ContainerProps} />;
  // Render
  return (
    <Container {...ContainerProps}>
      {each.map((item, i) => {
        // Map items
        const mapped = map(item, i);
        // Check falsy
        if (mapped) {
          // Render `Component`
          return <Component key={i} {...ComponentProps} {...mapped} />;
        }
        // Nothing to render
        return null;
      })}
    </Container>
  );
}

For.propTypes = {
  each: PropTypes.array,
  map: PropTypes.func,
  Component: PropTypes.elementType.isRequired,
  ComponentProps: PropTypes.object,
  Container: PropTypes.elementType,
  ContainerProps: PropTypes.object,
};
