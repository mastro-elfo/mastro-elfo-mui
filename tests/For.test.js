import React from "react";
import renderer from "react-test-renderer";

import For from "../src/components/For";

test("Render without props", () => {
  const component = renderer.create(<For Component={() => null} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Render null", () => {
  const component = renderer.create(<For each={null} Component={() => null} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Render undefined", () => {
  const component = renderer.create(
    <For each={undefined} Component={() => null} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Render empty", () => {
  const component = renderer.create(<For each={[]} Component={() => null} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Render", () => {
  const Component = ({ value, ...props }) => <span {...props}>{value}</span>;
  const Container = (props) => <div {...props} />;

  const component = renderer.create(
    <For
      each={Array(3).fill(true)}
      map={(_, i) => ({ value: i, "data-mapped-property": i })}
      Component={Component}
      Container={Container}
      ComponentProps={{ "data-prop": "component property" }}
      ContainerProps={{ "data-prop": "container property" }}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Filter falsy", () => {
  const component = renderer.create(
    <For
      each={[true, false, null, undefined, "", [], 0, 1]}
      Component={({ value }) => <span data-value={String(value)} />}
      map={(v) => !!v && { value: v }}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("No Component", () => {
  console.error = jest.fn();
  const component = renderer.create(<For />);
  expect(console.error).toHaveBeenCalledTimes(1);
});
