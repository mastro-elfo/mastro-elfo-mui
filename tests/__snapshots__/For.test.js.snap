// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter falsy 1`] = `
Array [
  <span
    data-value="true"
  />,
  <span
    data-value=""
  />,
  <span
    data-value="1"
  />,
]
`;

exports[`Render 1`] = `
<div
  data-prop="container property"
>
  <span
    data-mapped-property={0}
    data-prop="component property"
  >
    0
  </span>
  <span
    data-mapped-property={1}
    data-prop="component property"
  >
    1
  </span>
  <span
    data-mapped-property={2}
    data-prop="component property"
  >
    2
  </span>
</div>
`;

exports[`Render empty 1`] = `null`;

exports[`Render null 1`] = `null`;

exports[`Render undefined 1`] = `null`;

exports[`Render without props 1`] = `null`;
